Extending label.Parse method to support exact match

sdminonne opened the pull request:
@bgrant0607 @smarterclayton  I've a first implementation of the first step (2).
Not sure if you wish to have a look to this as a PR or you would like to
I modified labels.Parse to use an hand made recursive descendant parser. (hope it's not too much work to review it).
The previous parser was partially regexp based and it may become quickly complex to work with it.
At the same time I couldn't obtain more compact code since this "micro-language" is quite unusual:
for example it can parse something like this `x in (,)`.

Exact match is obtained using "in".
`x in (a)` produces the same matching of `x=a`, since internal data structure is the same.
String method will produce different output:
`Parse("x=a").String()` -> `"x=a"`
`Parse("x in (a)").String()` -> `"x in (a)"`.

In testing file:
Added labels.Parse in the requireMatch method to be sure (or at least for this tests) that the behavior is the same (where the requested code was not too complex).

Some modifications:
a) `a` (with trailing whitespace) now is matched and is valid. Skipping handling heading and trailing spaces in the exact way
b) `x not in ( )` is parsed and valid as `x in ( )`. (insure if I'm missing something)
c) `x not (,a)` (with leading comma) now is parsed and valid. Again if `x in (abc,)`  is parsable why it "x not in (,a)" should be equally parsable?
d) I'm still waiting for the @thockin and @smarterclayton about 'not in' vs "NotIn" and about capitalization (not a big deal to modify)

sdminonne committed:
Extending label.Parse method to support exact match

thockin commented on pkg/labels/selector.go:
```diff
@@ -303,24 +692,19 @@ func (lsel *LabelSelector) String() (string, error) {
-//      \s+
-//
+// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
+// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
```
Might be simpler to define it as we do in validation (much tighter spec)?

sdminonne replied:
I tried to BNF style I found. So would you prefer a regex style based as in validation.go?

thockin replied:
I'm fine with regex for simple token-scope docs and validation.  Obviously not for the full parser :)

thockin commented on pkg/labels/selector.go:
```diff
@@ -303,24 +692,19 @@ func (lsel *LabelSelector) String() (string, error) {
-//
+// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
+// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
+// VALUE is a sequence of zero or more characters that does not contain: ('=', '!', '(', ')', ',')
```
We don't actually do any verification of values at the moment - do we need to?

sdminonne replied:
@thockin  right.
Actually the verification is 'built-in' in the operator of this func https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/labels/selector.go#L449 
But it's true that  I've introduced this limitation "a key or a value cannot contain ')' or '('.  I can fix this even if I'm not sure I like having `key=(value)`

thockin replied:
The question is larger than just this PR, and has been somewhat on-going.

@smarterclayton @bgrant0607 

Do we want to restrict label values?  If so, we should do that at input validation.  If not, we need to define syntax for handling things like embedded spaces.

smarterclayton replied:
Sigh.

Having complex values seems wrong - it limits the ability of clients to bound behavior.  Embedded spaces seems like encouraging English language constructs in values.  Most tagging and categorization systems have gotten away without spaces.

We *should* reserve the backslash character (or similar) in the grammar and define it as an escape character from day one (even if it's not useful immediately).

smarterclayton replied:
I think I was on board with Brian's arguments for limiting keys and values of labels.

thockin replied:
Which arguments were these?  Last I saw there was a hangup about labels and annotations being different, but I am increasingly convince they ARE different.

smarterclayton replied:
I can't find it now - maybe I'm remembering the last face to face where we talked about being strict on labels to make the UI easier and clearer to read.

Using the value of a label as an input to machine (for processing) seems to be what annotations are for.  Wanting to find things by arbitrary annotations seems like an argument for exposing search/query on annotations via some extra service.  Wanting to find things by clear, unambiguous groupings and arrangements is labels, and it's intended to bridge a gap between machines (which work on the groupings) and humans (which need to understand the groupings).  Long, complex label values will be hard to use in UIs and from the CLI (certainly anything require bash escaping, or involving quotes, is doomed).  Having spaces in label values (unless clearly defined now) does seem like it opens the doors to "label query injection attacks" in the future.  The options are LabelPreparedStatements, or a restrictive value syntax.  If labels are like tagging systems, then there's reasonable practical experience saying spaces don't make or break.

Those are the thoughts in my head right now.

----- Original Message -----
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ','
> > or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain
> > ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> 
> Which arguments were these?  Last I saw there was a hangup about labels and
> annotations being different, but I am increasingly convince they ARE
> different.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24871642

thockin replied:
I agree.  Propose we keep label keys as qualified-name and force label
values as [A-Za-z0-9_-.]* - what think?

On Tue, Feb 17, 2015 at 5:25 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/labels/selector.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486#discussion_r24872401>
> :
>
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ',' or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain: ('=', '!', '(', ')', ',')
>
> I can't find it now - maybe I'm remembering the last face to face where we
> talked about being strict on labels to make the UI easier and clearer to
> read. Using the value of a label as an input to machine (for processing)
> seems to be what annotations are for. Wanting to find things by arbitrary
> annotations seems like an argument for exposing search/query on annotations
> via some extra service. Wanting to find things by clear, unambiguous
> groupings and arrangements is labels, and it's intended to bridge a gap
> between machines (which work on the groupings) and humans (which need to
> understand the groupings). Long, complex label values will be hard to use
> in UIs and from the CLI (certainly anything require bash escaping, or
> involving quotes, is doomed). Having spaces in label values (unless clearly
> defined now) does seem like it opens the doors to "label query injection
> attacks" in the future. The options are LabelPreparedStatements, or a
> restrictive value syntax. If labels are like tagging systems, then there's
> reasonable practical experience saying spaces don't make or break. Those
> are the thoughts in my head right now.
> ... <#14b9a4a99acd65cb_>
> ----- Original Message -----
>  > // <value-set> ::= "(" <values> ")" > // <values> ::= VALUE | VALUE ","
> <values> > -// > -// KEY is a sequence of one or more characters that does
> not contain ',' > or ' ' > -// [^, ]+ > -// VALUE is a sequence of zero or
> more characters that does not contain > ',', ' ' or ')' > -// [^, )]* > -//
> WHITESPACE_OPT is a sequence of zero or more whitespace characters > -//
> \s* > -// WHITESPACE is a sequence of one or more whitespace characters >
> -// \s+ > -// > +// <exact-match-restriction> ::= ["="|"=="|"!="]
> VALUE_NOT_EMPTY > +// KEY is a sequence of one or more characters that does
> not contain: > ('=', '!', '(', ')', ',') > +// VALUE is a sequence of zero
> or more characters that does not contain: > ('=', '!', '(', ')', ',') Which
> arguments were these? Last I saw there was a hangup about labels and
> annotations being different, but I am increasingly convince they ARE
> different. --- Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24871642
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24872401>
> .
>

smarterclayton replied:
Are we going to want Unicode characters?


> On Feb 18, 2015, at 12:00 AM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/labels/selector.go:
> 
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ',' or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain: ('=', '!', '(', ')', ',')
> I agree. Propose we keep label keys as qualified-name and force label values as [A-Za-z0-9_-.]* - what think?
> …
> —
> Reply to this email directly or view it on GitHub.
>

thockin replied:
Cross that bridge when someone complains?

On Tue, Feb 17, 2015 at 9:31 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/labels/selector.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486#discussion_r24879872>
> :
>
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ',' or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain: ('=', '!', '(', ')', ',')
>
> Are we going to want Unicode characters?
> ... <#14b9b2b4f73e18a0_>
>  On Feb 18, 2015, at 12:00 AM, Tim Hockin <notifications@github.com>
> wrote: In pkg/labels/selector.go: > // <value-set> ::= "(" <values> ")" >
> // <values> ::= VALUE | VALUE "," <values> > -// > -// KEY is a sequence of
> one or more characters that does not contain ',' or ' ' > -// [^, ]+ > -//
> VALUE is a sequence of zero or more characters that does not contain ',', '
> ' or ')' > -// [^, )]* > -// WHITESPACE_OPT is a sequence of zero or more
> whitespace characters > -// \s* > -// WHITESPACE is a sequence of one or
> more whitespace characters > -// \s+ > -// > +// <exact-match-restriction>
> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY > +// KEY is a sequence of one or more
> characters that does not contain: ('=', '!', '(', ')', ',') > +// VALUE is
> a sequence of zero or more characters that does not contain: ('=', '!',
> '(', ')', ',') I agree. Propose we keep label keys as qualified-name and
> force label values as [A-Za-z0-9_-.]* - what think? ... -- Reply to this email
> directly or view it on GitHub.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24879872>
> .
>

smarterclayton replied:
----- Original Message -----
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ','
> > or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain
> > ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> 
> Cross that bridge when someone complains?
> 

Only possible risk is then we have to whitelist the unicode char set we allow which sucks.  I think it's solvable...

smarterclayton replied:
I'll abdicate both responsibility and authority on this - I think limited set is ok, we should reserve \ specifically for escaping in the future, but I don't have conviction on those items enough to push others that way.  I'm ok with Tim's setup.

----- Original Message -----
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ','
> > or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain
> > ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> 
> Cross that bridge when someone complains?
> 
> On Tue, Feb 17, 2015 at 9:31 PM, Clayton Coleman <notifications@github.com>
> wrote:
> 
> > In pkg/labels/selector.go
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486#discussion_r24879872>
> > :
> >
> > >  //           <value-set> ::= "(" <values> ")"
> > >  //              <values> ::= VALUE | VALUE "," <values>
> > > -//
> > > -// KEY is a sequence of one or more characters that does not contain ','
> > > or ' '
> > > -//      [^, ]+
> > > -// VALUE is a sequence of zero or more characters that does not contain
> > > ',', ' ' or ')'
> > > -//      [^, )]*
> > > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > > -//      \s*
> > > -// WHITESPACE is a sequence of one or more whitespace characters
> > > -//      \s+
> > > -//
> > > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > > +// KEY is a sequence of one or more characters that does not contain:
> > > ('=', '!', '(', ')', ',')
> > > +// VALUE is a sequence of zero or more characters that does not contain:
> > > ('=', '!', '(', ')', ',')
> >
> > Are we going to want Unicode characters?
> > ... <#14b9b2b4f73e18a0_>
> >  On Feb 18, 2015, at 12:00 AM, Tim Hockin <notifications@github.com>
> > wrote: In pkg/labels/selector.go: > // <value-set> ::= "(" <values> ")" >
> > // <values> ::= VALUE | VALUE "," <values> > -// > -// KEY is a sequence of
> > one or more characters that does not contain ',' or ' ' > -// [^, ]+ > -//
> > VALUE is a sequence of zero or more characters that does not contain ',', '
> > ' or ')' > -// [^, )]* > -// WHITESPACE_OPT is a sequence of zero or more
> > whitespace characters > -// \s* > -// WHITESPACE is a sequence of one or
> > more whitespace characters > -// \s+ > -// > +// <exact-match-restriction>
> > ::= ["="|"=="|"!="] VALUE_NOT_EMPTY > +// KEY is a sequence of one or more
> > characters that does not contain: ('=', '!', '(', ')', ',') > +// VALUE is
> > a sequence of zero or more characters that does not contain: ('=', '!',
> > '(', ')', ',') I agree. Propose we keep label keys as qualified-name and
> > force label values as [A-Za-z0-9_-.]* - what think? ... -- Reply to this
> > email
> > directly or view it on GitHub.
> >
> > --
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24879872>
> > .
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24882205

sdminonne replied:
@smarterclayton I'll code Tim's proposal. Thanks for the prompt reply

bgrant0607 replied:
Sorry -- too many days of meetings. Yes, we should restrict label values, ideally with the same set of restrictions as we use for something else in the system. We should not restrict annotation values. I'd like label and annotation keys to follow the same rules, but my broader general position is that we shouldn't introduce too many different character restrictions in the system as a whole. If something is different than everything else, there needs to be a VERY good reason.

thockin commented on pkg/labels/selector.go:
```diff
@@ -303,24 +692,19 @@ func (lsel *LabelSelector) String() (string, error) {
+// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
+// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
+// VALUE is a sequence of zero or more characters that does not contain: ('=', '!', '(', ')', ',')
+// VALUE_NOT_EMPYT is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
```
misspelled EMPTY

sdminonne replied:
sorry for this I'll fix it

thockin commented on pkg/labels/selector_test.go:
```diff
@@ -306,65 +493,89 @@ func TestRequirementLabelSelectorMatching(t *testing.T) {
+			getRequirement("a", Exists, nil, SetBasedMatch, t),
+		}}, true, true},
+		{"a in (xyz abc)", nil, false, false},  // no comma
+		{"a in (x,y,z,in)", nil, false, false}, // operator 'in' inside list of identifiers
```
given that the parser knows the parse state is "inside a list", is "in" really restricted?  This suggests less keywords, to me.  I'm worried that labels will be seen and used as labels, and not as some validated syntax that every user of the system needs to know.

```
"labels": {
    "mydomain.com/in-or-out": "in"
}
```

That seems totally reasonable to the naive eye.

sdminonne replied:
in the actual PR only `in` and `not` are keywords (in the usual sense). It's true that we can add a `context` info to remove this limitation as well. This will permit to specify something like:
`answer not in (not)` which looks funny but it makes sense.
OK.

thockin replied:
Before you burn a lot of time, let's get an answer around the overall value validation question.  It's been lingering long enough.

bgrant0607 replied:
`in` and `not` should be legal values.

thockin commented:
Needs testing somewhere for keys that fit the qualified name syntax for keys:  [ DNS_SUBDOMAIN "/" ] DNS_LABEL.

Specifically, test cases like "example.com/my-key" and "my-key-with-a-long-name" and "0.1.2.domain/99" are all valid as per validation rules today.

sdminonne commented:
@thockin I've add a very simple test here https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#diff-b2970db202ffdb1dec2b170d88679d3bR504

thockin commented on pkg/labels/selector.go:
```diff
@@ -333,114 +717,18 @@ func (lsel *LabelSelector) String() (string, error) {
 func validateLabelKey(k string) error {
-	if !util.IsDNSLabel(k) {
+	if !util.IsQualifiedName(k) {
 		return errors.NewFieldNotSupported("key", k)
```
This should be NewFieldInvalid()

thockin commented on pkg/labels/selector_test.go:
```diff
@@ -306,65 +493,89 @@ func TestRequirementLabelSelectorMatching(t *testing.T) {
-			getRequirement("x", Exists, nil, t),
+			getRequirement("x", Exists, nil, SetBasedMatch, t),
+		}}, true, true},
+		{"this-is-a-dns.domain.com/key-with-dash", &LabelSelector{Requirements: []Requirement{
```
We should also cover this case in the fuller expression evaluator, e.g. "this-is-a-dns.domain.com/key-with-dash in (so,what)".

We should also have coverage for keys that look like numbers (which are actually valid).  A key called "0" is legit, I think.

sdminonne commented:
@thockin @smarterclayton in the next hours I will submit and update to fix the less controversial issues (typo, testcases, error report function)
For the regex to validate the value I don't think is going to be an issue (except if we move to bytecode). In the meantime if you converge to something like `[A-Za-z0-9_-.]*` just let me know.

bgrant0607 commented:
On `not in` vs. NotIn:

Our general pattern is that constants in API fields are CameCase, like `NotIn` (e.g., `ClusterFirst`), and that constants expressed as part of the URL are lowercase versions of those, with no spaces or punctuation, such as `notin` (e.g., `replicationcontrollers`). I recommend we follow that same pattern here.

WDYT, @smarterclayton ?

sdminonne commented:
@smarterclayton @thockin @bgrant0607 
Fixed the TYPO
'in' and 'notin' now can be value of the label (modified)
Modified 'not in'  to 'notin'
Added the regexp suggested by @thockin + `\` char for future uses as suggested by @smarterclayton . (Addes some validation tests as well)
Hope it's better this time

bgrant0607 commented on pkg/labels/selector_test.go:
```diff
@@ -306,65 +501,97 @@ func TestRequirementLabelSelectorMatching(t *testing.T) {
+		{"x=a", &LabelSelector{Requirements: []Requirement{
+			getRequirement("x", In, util.NewStringSet("a"), ExactMatch, t),
+		}}, true, true},
+		{"x=a,y!=b", &LabelSelector{Requirements: []Requirement{
```
I would like to see a test that combines the exact-match syntax and the set-based syntax, such as:
```
x=a,y!=b,z in (h,i,j)
```

sdminonne replied:
OK.

bgrant0607 commented on pkg/labels/selector.go:
```diff
@@ -162,21 +150,34 @@ const (
+// Requirement implements both set based match and exact match
+// Requirement is initialized via NewRequirement constructor for creating a valid Requirement.
 type Requirement struct {
+	matchType MatchType
```
We're keeping track of MatchType so that we can serialize the selector in the same form it was provided? Why not just use Operator to distinguish them?

bgrant0607 replied:
Don't feel obligated to change this. It's an internal data structure, and making MatchType explicit is pretty clear, so I don't care too much.

sdminonne replied:
Yes MatchType is there to serialize selector in 'almost' the same way.
`x==a` is going to be serialized in `x=a`. I wouldn't consider this a bug but someone else could disagree. 
On the other side the serialization is not the verbatim input since we `sort` the data structure to achieve determinism. I don't know. I'm OK to change it

smarterclayton commented:
This is purely a personal opinion - but I think NotIn would be ugly in general use.

Seeing "foo not in bar" or "fo in bar" seems much clearer than "foo In bar" and "foo NotIn bar".  I generally prefer lower case for operator words (or caseless)  before mixed case.  Again, this purely an "attractiveness" argument.


> On Feb 18, 2015, at 9:34 PM, Brian Grant <notifications@github.com> wrote:
> 
> On not in vs. NotIn:
> 
> Our general pattern is that constants in API fields are CameCase, like NotIn (e.g., ClusterFirst), and that constants expressed as part of the URL are lowercase versions of those, with no spaces or punctuation, such as notin (e.g., replicationcontrollers). I recommend we follow that same pattern here.
> 
> WDYT, @smarterclayton ?
> 
> —
> Reply to this email directly or view it on GitHub.
>

thockin commented:
yeah, NotIn is hurtful to me.  I don't think an embedded grammar syntax
needs to follow the same rules for constants.  I'd just as soon see x IN y
and x NOTIN y or something.

What about !in or !(x in y)?  We have to decide if this just a small hacky
thing or an actual grammar.  Do we support parens (I think no), for example.

As for keys and values:  I TOTALLY agree that reusing common rules is a
good thing.  I almost can't keep straight the ones we have much less new
ones.

We need to have agreement on:

* label values: restrict to [A-Za-z0-9_-.]*
* annotation values: no restriction

Now for keys.  We could just loosen all qualified names to be
([A-Za-z0-9_-.]+/)?[A-Za-z0-9_-.]+ - this should be a strict superset of
what we allow today.  We can say that convention is to use dns-compatible
domain + single token, but still allow things like FOO/B_A_R.  This would
impact resource names and label keys.

Alternately we could define it as caseless - but that impacts all callers.

What think?

On Thu, Feb 19, 2015 at 10:16 AM, Clayton Coleman <notifications@github.com>
wrote:

> This is purely a personal opinion - but I think NotIn would be ugly in
> general use.
>
> Seeing "foo not in bar" or "fo in bar" seems much clearer than "foo In
> bar" and "foo NotIn bar". I generally prefer lower case for operator words
> (or caseless) before mixed case. Again, this purely an "attractiveness"
> argument.
>
>
> > On Feb 18, 2015, at 9:34 PM, Brian Grant <notifications@github.com>
> wrote:
> >
> > On not in vs. NotIn:
> >
> > Our general pattern is that constants in API fields are CameCase, like
> NotIn (e.g., ClusterFirst), and that constants expressed as part of the URL
> are lowercase versions of those, with no spaces or punctuation, such as
> notin (e.g., replicationcontrollers). I recommend we follow that same
> pattern here.
> >
> > WDYT, @smarterclayton ?
> >
> > --
> > Reply to this email directly or view it on GitHub.
> >
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486#issuecomment-75105534>
> .
>

sdminonne commented:
Added the mixed test: `x=a,y!=b,z in (h,i,j)`.
Removed `MatchType` to simplify interface for `NewRequirement` (price is more complex `String()` and `parseOperator`).

thockin commented on pkg/labels/selector.go:
why do we support and distinguish = and == both?

thockin commented:
I don't have any further gripes

bgrant0607 commented:
Re. `NotIn`: That would just be for the API struct representation. There will be some Operator field. You're saying that the values shouldn't follow our rules about constants in the API?

thockin commented:
Ah, I meant the stringified syntax being "NotIn" is painful.  Go-level
constants is fine :)

On Fri, Feb 20, 2015 at 11:03 AM, Brian Grant <notifications@github.com>
wrote:

> Re. NotIn: That would just be for the API struct representation. There
> will be some Operator field. You're saying that the values shouldn't follow
> our rules about constants in the API?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486#issuecomment-75298952>
> .
>

bgrant0607 commented:
To be clear, it would be an API constant string, not just a Go type. See above comparison with `ClusterFirst`.

As for URL syntax, the decision is `notin` vs `not in`. The former is arguably more consistent. The latter is more readable. I initially proposed the latter, then suggested the former. I don't really have a strong opinion, I guess. In the URL, the space will need to be escaped, so `not%20in` vs `notin`.

sdminonne commented:
@bgrant0607  modified operators to be constant strings. Sorry for the delay.

sdminonne commented:
@bgrant0607 I made a mess with my previous push, sorry hope you hadn't look at it.

bgrant0607 commented:
@sdminonne is this ready to look at now?

sdminonne commented:
@bgrant0607 it is. Thanks

bgrant0607 commented:
LGTM for this round.

Please rebase so tests can run.

@thockin @smarterclayton Any more comments on this PR?

smarterclayton commented on pkg/labels/selector.go:
This is a nit (just a style thing for future pulls): go encourages using a switch whenever you have this sort of pattern - it ends up reading better.  Basically go hates "else if" and wants switch.  If you ever are in here again later it's a minor thing to clean up if you can.

smarterclayton commented:
LGTM

sdminonne commented on pkg/labels/selector.go:
@smarterclayton  I'm aware of this "go preferred way" but I've to get rid of some old habits. I'll change it. Thanks.

sdminonne commented:
@smarterclayton @bgrant0607 I'm going to rebase and  but all my PRs are always  failing for Shippable. And I trying to get info via

sdminonne commented:
QED :)
![image](https://cloud.githubusercontent.com/assets/9527409/6367657/c178ce56-bcd5-11e4-9364-d3d2a47dd313.png)

bgrant0607 commented:
Integration test failed due to a pod not scheduling before the timeout, but only for Go 1.3. Probably a flake.
